### El problema de los misioneros y los caníbales en AIMA

In [2]:
# importamos las cosas que vamos a usar de aima
from search import *

class ProblemaMisioneros(Problem):
    ''' Clase problema (formalizacion de nuestro problema) siguiendo la
        estructura que aima espera que tengan los problemas.'''
    def __init__(self, initial, goal=None):
        '''Inicializacion de nuestro problema.'''
        Problem.__init__(self, initial, goal)
        # cada accion tiene un texto para identificar al operador y despues una tupla con la
        # cantidad de misioneros y canibales que se mueven en la canoa
        self._actions = [('1c', (0,1)), ('1m', (1, 0)), ('2c', (0, 2)), ('2m', (2, 0)), ('1m1c', (1, 1))]

    def actions(self, s):
        '''Devuelve las acciones validas para un estado.'''
        # las acciones validas para un estado son aquellas que al aplicarse
        # nos dejan en otro estado valido
        return [a for a in self._actions if self._is_valid(self.result(s, a))]

    def _is_valid(self, s):
        '''Determina si un estado es valido o no.'''
        # un estado es valido si no hay mas canibales que misioneros en ninguna
        # orilla, y si las cantidades estan entre 0 y 3
        return (s[0] >= s[1] or s[0] == 0) and ((3 - s[0]) >= (3 - s[1]) or s[0] == 3) and (0 <= s[0] <= 3) and (0 <= s[1] <= 3)

    def result(self, s, a):
        '''Devuelve el estado resultante de aplicar una accion a un estado
           determinado.'''
        # el estado resultante tiene la canoa en el lado opuesto, y con las
        # cantidades de misioneros y canibales actualizadas segun la cantidad
        # que viajaron en la canoa
        if s[2] == 0:
            return (s[0] - a[1][0], s[1] - a[1][1], 1)
        else:
            return (s[0] + a[1][0], s[1] + a[1][1], 0)



In [9]:
# creamos un problema a partir de nuestra formalizacion de ProblemaMisioneros
# como parametros le pasamos el estado inicial, y el estado meta que esperamos
estado = ProblemaMisioneros((3, 3, 0), (0, 0, 1))

# le decimos a aima que resuelva nuestro problema con el metodo de busqueda en
# amplitud
breadth_first_tree_search(estado).solution()


[('2c', (0, 2)),
 ('1c', (0, 1)),
 ('2c', (0, 2)),
 ('1c', (0, 1)),
 ('2m', (2, 0)),
 ('1m1c', (1, 1)),
 ('2m', (2, 0)),
 ('1c', (0, 1)),
 ('2c', (0, 2)),
 ('1c', (0, 1)),
 ('2c', (0, 2))]

In [12]:
%%timeit
breadth_first_tree_search(estado).solution()

104 ms ± 8.18 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


### Ejercicio opcional. Define alguna heurística y estudia las propiedades del algoritmo A*

In [4]:
def h1(node):
    '''
        Devuelve 2 veces el numero de personas a la izquierda y
        si la barca esta a la izquierda la resta 1.
    '''
    goal = (0, 0, 1)
    misioneros_izq = node.state[0]
    canibales_izq = node.state[1]
    pos_barca = node.state[2]
    return 2 * (misioneros_izq + canibales_izq) - pos_barca

In [10]:
astar_search(estado,h1).solution()

[('1m1c', (1, 1)),
 ('1m', (1, 0)),
 ('2c', (0, 2)),
 ('1c', (0, 1)),
 ('2m', (2, 0)),
 ('1m1c', (1, 1)),
 ('2m', (2, 0)),
 ('1c', (0, 1)),
 ('2c', (0, 2)),
 ('1c', (0, 1)),
 ('2c', (0, 2))]

In [11]:
%%timeit
astar_search(estado,h1).solution()

187 µs ± 20.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


#### Respuesta:

Como podemos observar si definimos una heuristica admisible podemos ver que el tiempo de ejecucion del algoritmo A* es inferior al de busqueda ciega en anchura sin control de repetidos, quedando los tiempos de la siguiente manera:

<center>
    
|Algoritmo|Tiempo|
|---------|------|
|Busqueda Ciega en Anchura|104 ms|
|A*|187 µs|

</center>
